### PR TITLE
2094 Offset (skip-take) Paging. Out of the box support

### PR DESCRIPTION
**Pull request by ademchenko**
2094 Offset (skip-take) Paging. Out of the box support

- done

Addresses #2094 (in this specific format)

### CODE REVIEW

**Comment by michaelstaib** [comment]
hi @ademchenko,

thanks for taking the time and submitting this PR.

We will have a look at it and then come back with comments on this.

**Comment by ademchenko** [comment]
> hi @ademchenko,
> 
> thanks for taking the time and submitting this PR.
> 
> We will have a look at it and then come back with comments on this.

Hi, @michaelstaib , looking forward to getting the review done and hope the change would be worth enough to be merged into the Hotchocolate repository.

**Comment by michaelstaib** [comment]
@ademchenko, we think that this paging will be moved into a separate project, not into the type core. In general offset, paging is something we want to do but we are at the moment busy getting the new execution engine and data abstractions together. Only after this, we will have a look at your pr.

So this, for now, has to wait for a little. we will I think come back to this in two or three weeks.

Sorry for the wait but these two changes are massive and really demand all our attention.

File: src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs:5]: @ademchenko I am not happy with the name CollectionSlice we may should look for a different one.

@PascalSenn @rstaib do we have better name suggestions?

File: src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs:9]: @ademchenko shall we move this into a separate interface `IPageInfo` like with the connection base paging?

**Comment by michaelstaib** [comment]
@ademchenko good start on this one. Sorry for the wait, we did quite a lot of changes for 11 and wanted to wait until we hade more code merged to master in order to not run into big merge conflicts. I think we should move this into a separate assembly `HotChocolate.Types.OffsetPaging`. We will soon do the same with the connection base paging. We will reference this package in the meta-package `HotChocolate` so that everybody will automatically get it.

I will do some style changes over the next couple of days to the code so that it fits in with the rest of the code. We have a 100 character rule.

@PascalSenn @rstaib can you also have a look over this?

**Comment by michaelstaib** [comment]
OK, the changes that we merged in introduced compile errors in your code... I will have a look at it.

File: src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs
Comment (PascalSenn) [reply src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs:5]: Maybe:
ICollectionPartition
ICollectionChunk
ICollectionSegement
ICollectionSector

Maybe event withouth the Collection part
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq/src/System/Linq/IPartition.cs

Comment (ademchenko) [reply src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs:5]: I think ICollectionSegment is a good choice since it is consistent with .net naming ArraySegment. At least it should be clear for library users.

File: src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs
Comment (ademchenko) [reply src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs:9]: @michaelstaib, I didn't do that because that IPageInfo would contain just one property TotalCount instead of five in the case of cursor-based paging. Actually I don't think this is subject to change in the future so could we leave as is?

File: src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs:5]: Or partition sounds also good.

**Comment by michaelstaib** [comment]
We have one more PR to merge before we are including this into the version 11 branch ...

**Comment by michaelstaib** [comment]
@ademchenko @PascalSenn I have started to integrate it ... It should become part of the next preview release probably tomorrow.

File: src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs]: @ademchenko we are going with the segment.

File: src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/Types/OffsetPaging/ICollectionSlice.cs]: The question is if we want to also introduce some indicator props that give us more info about the current segment like HasNextPage / HasPrevious page.

Anyhow, we will make it nullable since the `ICollectionSegmentResolver`can choose to not allow `totalCount` for performance reasons. The default will be to not have a total count. We also reworked the cursor pagination this way.

File: src/HotChocolate/Core/src/Types/Types/OffsetPaging/OffsetPagingObjectFieldDescriptorExtensions.cs
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/Types/OffsetPaging/OffsetPagingObjectFieldDescriptorExtensions.cs]: We have reworked the type system and this should no longer pose an issue... you can have a look once I am done with refactoring ....

**Comment by michaelstaib** [comment]
- [ ] Introduce new paging abstraction
- [ ] Integrate basic cursor paging
- [ ] add cursor paging tests
- [ ] Integrate basic offset paging
- [ ] add offset paging tests
- [ ] add xml docs
- [ ] create docs

**Comment by michaelstaib** [comment]
@ademchenko with version 11 SchemaTypeResolver is now public as well as a couple of other new types that. We have implemented the new filter api without access to internals which was not possible with 10.

**Comment by michaelstaib** [comment]
With the latest check-in, the basics are in ... also we have now proper integration tests. Will reintegrate the cursor pagination now and move tests to the new test projects. After that, we will start with further refinements. Should be merged today.

**Review by michaelstaib — APPROVED** [review]

**Pull request merged**
